iceoryx_examples/ice_multi_publisher/ice_re_subscriber.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/popo/guard_condition.hpp"
+#include "iceoryx_posh/popo/modern_api/typed_subscriber.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "topic_data.hpp"
+
+#include <chrono>
+#include <csignal>
+#include <iostream>
+
+bool killswitch = false;
+
+static void sigHandler(int sig [[gnu::unused]])
+{
+    killswitch = true;
+}
+
+constexpr uint64_t MAX_HISTORY_SIZE{4};
+constexpr uint64_t UNSUBSCRIBED_TIME_SECONDS{3};
+
+void receive()
+{
+    iox::popo::TypedSubscriber<CounterTopic> subscriber({"CounterTopic", iox::capro::AnyInstanceString, "Counter"});
+
+    subscriber.subscribe();
+    uint64_t maxNumSamples = MAX_HISTORY_SIZE - 2;
+    while (!killswitch)
+    {
+        // unsubscribe and resubscribe
+        subscriber.unsubscribe();
+        std::cout << "Unsubscribed ... Resubscribe in " << UNSUBSCRIBED_TIME_SECONDS << " seconds" << std::endl;
+
+        // we will probably miss some data while unsubscribed
+        std::this_thread::sleep_for(std::chrono::seconds(UNSUBSCRIBED_TIME_SECONDS));
+
+        // we (re)subscribe with differing maximum number of samples
+        // and should see at most the latest last maxNumSamples
+        maxNumSamples = maxNumSamples % MAX_HISTORY_SIZE + 1; // cycles between last 1 to MAX_HISTORY_SIZE samples
```

Comment:
We have to discuss history and queue size. You change the queue size here. History is what the publisher builds for last sent samples and a subscriber can get on subscribe. But you did not set it for publisher and subscriber. This all belongs to a proper API description. Here it is getting confusing

Reply:
Right. This is the number of samples the subscriber expects. I verified that it can be changed dynamically (the API allows this kind of resubscribe) but it must be compatible to the history the publisher keeps, i.e. if we expect k samples on subscriber side the history size at publisher side must be at least k as well.

Reply:
But you did not provide a history for the publisher. I.e. it is the default size of 0. Your queue can store the samples but when subscribing you don't get the missed ones. I'm not sure if you only test the queuing capabilities here or if you also need the history. When it's just about the queue, you should talk about queue size and not history size

Reply:
Is this even possible with the typed API?

Reply:
No that seems to be missing :-(